action/protocol/staking/candidate_center.go

```diff
@@ -7,143 +7,432 @@
 		ownerMap         map[string]*Candidate
 		operatorMap      map[string]*Candidate
 		selfStkBucketMap map[uint64]*Candidate
 	}
+
+	// CandidateCenter is a struct to manage the candidates
+	CandidateCenter struct {
+		*candBase
+		size   int
+		change CandidateView
+	}
 )
 
+// NewCandidateView creates an instance of CandidateView
+func NewCandidateView(l CandidateList) (CandidateView, error) {
+	cv := candChange{
+		dirty: make(map[string]*Candidate),
+	}
+
+	for _, d := range l {
+		if err := d.Validate(); err != nil {
+			return nil, err
+		}
+		cv.dirty[d.Owner.String()] = d
+	}
+	return &cv, nil
+}
+
 // NewCandidateCenter creates an instance of CandidateCenter
-func NewCandidateCenter() *CandidateCenter {
+func NewCandidateCenter() CandidateControl {
 	return &CandidateCenter{
-		nameMap:          make(map[string]*Candidate),
-		ownerMap:         make(map[string]*Candidate),
-		operatorMap:      make(map[string]*Candidate),
-		selfStkBucketMap: make(map[uint64]*Candidate),
+		candBase: &candBase{
+			nameMap:          make(map[string]*Candidate),
+			ownerMap:         make(map[string]*Candidate),
+			operatorMap:      make(map[string]*Candidate),
+			selfStkBucketMap: make(map[uint64]*Candidate),
+		},
+		change: newCandChange(),
 	}
 }
 
 // Size returns number of candidates
-func (m CandidateCenter) Size() int {
-	return len(m.nameMap)
+func (m *CandidateCenter) Size() int {
+	return m.size
 }
 
 // All returns all candidates in candidate center
-func (m CandidateCenter) All() (CandidateList, error) {
-	list := make(CandidateList, 0, m.Size())
-	for _, c := range m.nameMap {
-		list = append(list, c)
+func (m *CandidateCenter) All() CandidateList {
+	list := m.change.All()
+	if list == nil {
+		list = make(CandidateList, 0, len(m.ownerMap))
+	}
+
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	for _, d := range m.ownerMap {
+		if !m.change.ContainsOwner(d.Owner) {
+			list = append(list, d.Clone())
+		}
+	}
+	return list
+}
+
+// Base returns the confirmed base state
+func (m CandidateCenter) Base() CandidateControl {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return &CandidateCenter{
+		candBase: m.candBase,
+		size:     len(m.candBase.ownerMap),
+		change:   newCandChange(),
+	}
+}
+
+// Delta exports the pending changes
+func (m *CandidateCenter) Delta() CandidateList {
+	return m.change.All()
+}
+
+// SetDelta sets the delta
+func (m *CandidateCenter) SetDelta(c CandidateList) error {
+	m.change = nil
+	m.change = newCandChange()
+
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	m.size = len(m.ownerMap)
+	for _, v := range c {
+		if err := m.change.Upsert(v); err != nil {
+			return err
+		}
+		if _, ok := m.ownerMap[v.Owner.String()]; !ok {
+			m.size++
+		}
 	}
-	return list, nil
+	return nil
+}
+
+// Commit writes the list into center
+func (m *CandidateCenter) Commit(c CandidateList) error {
+	if len(c) == 0 {
+		// nothing to commit, simply return
+		return nil
+	}
+
+	m.change = nil
+	m.change = newCandChange()
```

Comment:
1. since "change" is in Candidate Center, why pass in "c"?
2. even we need "c", len(c) is 0 or not has different behavior, which is not expected.

Reply:
"change" is the final confirmed change, it comes from workingset's Dock

i can reformat it to a call of SetDelta(), then Commit(), this would be easier to understand